Expose theme_supports data through a minimial theme controller

desrosj opened the pull request:
## Description
This PR moves the theme support data previously in the REST API index (landed in #6318) into a read-only theme controller for the active theme.

I went with `wp/v2/themes/active` instead of `current` to match the wording in wp-admin and the codebase. I also went with the `edit_posts` capability, as anyone using the edit screen will need Gutenberg to know what the active theme supports.

I included the theme name, stylesheet (directory name), version, template, and theme supports data (what Gutenberg needs). I don't feel strongly about including any of those fields, but those are a few straightforward properties of a `WP_Theme` object that I felt were ok to include.

This is the first step for landing [Trac 45016](https://core.trac.wordpress.org/ticket/45016) in Core.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

desrosj committed:
First pass at adding a Themes endpoint for the current active theme.

desrosj committed:
Don't return `false` when the theme does not support thumbnails. Only list view context on formats and post-thumbnails fields. Uncomment permissions logic. Remove old code placing this data in the index.

desrosj committed:
Use the new active theme endpoint for determining theme support.

desrosj committed:
Add unit tests for the new active theme controller.

youknowriad commented on packages/core-data/src/resolvers.js:
```diff
@@ -70,8 +70,8 @@ export function* getEntityRecords( kind, name, query = {} ) {
 export function* getThemeSupports() {
-	const index = yield apiFetch( { path: '/' } );
-	yield receiveThemeSupportsFromIndex( index );
+	const currentTheme = yield apiFetch( { path: '/wp/v2/themes/active' } );
```
Maybe we can preload this API endpoint here https://github.com/WordPress/gutenberg/blob/master/lib/client-assets.php#L1303 ?

desrosj replied:
Added this endpoint to that list. It correctly prevents the Featured Image editor from flashing onto the screen like it was.

youknowriad replied:
Cool, do we need to pass the `edit` context here? (and the preloaded path)?

desrosj replied:
That I am not sure of. The default context is now `edit` because that is the only context this endpoint currently supports. But might be good to include it for when/if that default changes.

desrosj replied:
Context was removed entirely [per the discussion in #core-restapi](https://wordpress.slack.com/archives/C02RQC26G/p1539297122000100).

danielbachhuber commented on lib/class-wp-rest-themes-controller.php:
```diff
@@ -0,0 +1,233 @@
+					),
+				),
+			),
+		);
```
Each of these schema attributes should be `context=edit`, as Theme objects are only accessed in the Edit context.

danielbachhuber reviewed (changes requested):
Great start! Few more minor issues and then it should be good to go.

danielbachhuber commented:
Flagged @kadamwhite to review too, only if he has any opinions on this.

desrosj committed:
Pre-load the active theme endpoint.

desrosj committed:
PR Review Feedback adjustments:
- Alignment fixes
- Add missing textdomain arguments
- Update the property contexts to edit (should be the context the data is accessed)
- Add missing description for `post-thumbnails`
- Link should be `self`, not `active`

desrosj commented:
Made adjustments suggested. Also corrected the indentation and missing text domain errors picked up by PHPCS.

desrosj committed:
Fix failing tests. Change default context to edit.

desrosj committed:
Add missing textdomain.

danielbachhuber commented on lib/load.php:
```diff
@@ -15,6 +15,7 @@
 	require dirname( __FILE__ ) . '/class-wp-rest-blocks-controller.php';
 	require dirname( __FILE__ ) . '/class-wp-rest-autosaves-controller.php';
 	require dirname( __FILE__ ) . '/class-wp-rest-block-renderer-controller.php';
+	require dirname( __FILE__ ) . '/class-wp-rest-themes-controller.php';
```
We need to wrap this with a `class_exists()` but it can happen in a follow-up PR for all of the controllers: #10526

danielbachhuber reviewed (approved):
👍 Looks to me

desrosj committed:
Remove contexts from themes (see https://wordpress.slack.com/archives/C02RQC26G/p1539297122000100).

danielbachhuber commented:
@desrosj While you're at it, can you merge `master` for #10531 and apply the `class_exists()` check?

TimothyBJacobs commented:
I don't think the magic `__get` accessor should be used directly instead either just the property syntax, `$theme->name` or `$theme->get( 'Name' )` 

Additionally, `WP_Theme` has a display method for returning the "display" form of a lot of that data. If we include things like the theme name and such, we should probably include that and follow the same format we do elsewhere.

```
$data['name'] = array(
	'raw'      => $theme->get( 'Name' ),
	'rendered' => $theme->display( 'Name' )
);
```

If we do do that, we should then add the context property back and only expose `raw` on `edit` context. Or just omit the raw version entirely.

TimothyBJacobs commented on lib/class-wp-rest-themes-controller.php:
```diff
@@ -0,0 +1,221 @@
+
+			$post_thumbnails = get_theme_support( 'post-thumbnails' );
+
+			if ( $post_thumbnails ) {
```
Should we always include `post-thumbnails` in the `theme_supports` and set it to `false`?

kadamwhite replied:
I’d think so; it’s good to be able to predict the shape of our responses, consistently including properties is useful

desrosj replied:
The changes now ensure `post-thumbnails` is always included in `theme_supports`. The name, etc. fields are now removed.

desrosj committed:
Merge branch 'master' of github.com:WordPress/gutenberg into add/core-rest-api-themes-controller

# Conflicts:
#	lib/load.php

desrosj committed:
Always include `post-thumbnails` property to ensure consistent responses.

desrosj committed:
Avoid using the `__get()` magic method.

kadamwhite commented on lib/class-wp-rest-themes-controller.php:
```diff
@@ -0,0 +1,222 @@
+	public function register_routes() {
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base . '/active',
```
Per [discussion in Slack](https://wordpress.slack.com/archives/C02RQC26G/p1539359220000100), we should use the following structure for querying this information:
- `/wp/v2/themes?status=active` will return an array containing the object representing the capabilities of the active theme.
- For this initial release, `status` will be an `enum` parameter with one option, `active`, and that parameter will be required.
  - This leaves us room to expand the functionality of the endpoint to permit listing all themes down the road, etc, but restricts the complexity to the bare minimum currently needed.
  - Using a filter further avoids any lock-in around how to point to the active theme from other resources (the debate between a link from the index, `/active`, `/~active`, etc, which has unfolded mostly in slack). Any of these approaches might be found to be limiting down the road as we add support for theme management, multi-site, _etcetera_, while a query parameter like `status` is unlikely to conflict with any future plans.

Long-term I imagine we'd update the active theme by PUT'ing a theme slug to the settings endpoint, or a hypothetical new site endpoint. Figuring all that out is out of scope and not necessary for Gutenberg in terms of 5.0.

kadamwhite commented on lib/class-wp-rest-themes-controller.php:
```diff
@@ -0,0 +1,222 @@
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function get_item( $request ) {
```
Per the comment about `?status=active`, it would probably be best to introduce a `get_items` method since we will technically be querying the collection endpoint to return the active theme.

kadamwhite commented on lib/class-wp-rest-themes-controller.php:
```diff
@@ -0,0 +1,222 @@
+		}
+
+		if ( in_array( 'version', $fields, true ) ) {
+			$data['version'] = $theme->get( 'Version' );
```
Are stylesheet, version, etc strictly required for Gutenberg? No argument that they seem like they would be useful long-term but I feel it best to limit what we expose on this endpoint to the bare minimum of theme-supports values we know we need now, and add more later.

desrosj committed:
Change the `themes/active` route to `themes?status=active`. The former would cause future issues when `themes/themename` returns data for a specific theme.

For now, limit the `status` param to one value, `active`, by specifying the param as an `enum`.

Also, remove field not 100% required for Gutenberg (name, version, template, etc.).

desrosj committed:
Change variable name from `currentTheme` to `activeThemes`, as the endpoint now returns a collection.

danielbachhuber commented on lib/class-wp-rest-themes-controller.php:
```diff
@@ -0,0 +1,251 @@
+	protected function prepare_links( $theme ) {
+		$links = array(
+			'self' => array(
+				'href' => rest_url( sprintf( '%s/%s?status=active', $this->namespace, $this->rest_base ) ),
```
I think `href` is supposed to be `/wp/v2/themes/<slug>`, which we don't have right now. We could probably remove `prepare_links()` entirely.

danielbachhuber commented on lib/class-wp-rest-themes-controller.php:
```diff
@@ -0,0 +1,251 @@
+			'type'       => 'object',
+			'properties' => array(
+				'theme_supports' => array(
+					'description' => __( 'A list of features this theme supports.', 'gutenberg' ),
```
Language tweak: `Features supported by this theme.`

danielbachhuber reviewed (changes requested):
Looking good! Just a couple small nits on my end.

desrosj committed:
Remove `_links` from the response and adjust the theme supports property description.

desrosj committed:
Merge branch 'master' of github.com:WordPress/gutenberg into add/core-rest-api-themes-controller

kadamwhite reviewed (approved):
Changes look good, thank you @desrosj ! Only outstanding question would be whether we should add a unit test assertion for themes which do not support post formats, to validate that we get that empty array.

desrosj committed:
Updating JavaScript documentation.

desrosj commented:
@kadamwhite I believe `test_theme_supports_formats()` will verify that scenario. The code here will always return at least `array( 'standard' )`. Do you think `standard` should be excluded if there is no post format support?

kadamwhite commented:
> @kadamwhite I believe `test_theme_supports_formats()` will verify that scenario. The code here will always return at least `array( 'standard' )`. Do you think `standard` should be excluded if there is no post format support?

No opinion, @desrosj ; :+1: on the current test sufficiently covering that use-case.

danielbachhuber committed:
Rename `TotalThemes` to `TotalPages`
